rudder-core/src/main/scala/com/normation/rudder/services/reports/ReportingServiceImpl.scala

```diff
@@ -85,6 +106,143 @@ class ReportingServiceImpl(
     }
   }
 
+}
+
+trait CachedFindRuleNodeStatusReports extends ReportingService with CachedRepository with Loggable {
+
+  /**
+   * underlying service that will provide the computation logic
+   */
+  def defaultFindRuleNodeStatusReports: DefaultFindRuleNodeStatusReports
+  def nodeInfoService                 : NodeInfoService
+
+  /**
+   * The cache is managed node by node.
+   * A missing nodeId mean that the cache wasn't initialized for
+   * that node.
+   */
+  private[this] var cache = Map.empty[NodeId, Set[RuleNodeStatusReport]]
+
+  private[this] def cacheToLog(c: Map[NodeId, Set[RuleNodeStatusReport]]): String = {
+    //display compliance value and expiration date.
+    c.map { case (nodeId, reports) =>
+
+      val reportsString = reports.map { r =>
+        val run = r.agentRunTime match {
+          case None => "no agent run"
+          case Some(r) => s"last run@${r}"
+        }
+        val cfid = r.configId match {
+          case None => "no config id"
+          case Some(c) => s"configId:${c.value}"
+        }
+        s"${r.ruleId.value}/${r.serial}[exp:${r.expirationDate}][${run}][${cfid}]${r.compliance.toString}"
+      }.mkString("\n  ", "\n  ", "")
+
+      s"node: ${nodeId.value}${reportsString}"
+    }.mkString("\n", "\n", "")
+  }
+
+
+  /**
+   * Invalidate some keys in the cache. That won't charge them again
+   * immediately
+   */
+  def invalidate(nodeIds: Set[NodeId]): Unit = this.synchronized {
+    logger.debug(s"Compliance cache: invalidate cache for nodes: [${nodeIds.map { _.value }.mkString(",")}]")
+    cache = cache -- nodeIds
```

Comment:
shouldn't it be --= ?
http://www.scala-lang.org/api/current/#scala.collection.mutable.Map says
 --=(xs: TraversableOnce[A]): Map.this.type
Removes all elements produced by an iterator from this shrinkable collection.


 --(xs: GenTraversableOnce[A]): Map[A, B]
Creates a new map with all the key/value mappings of this map except mappings with keys equal to any of those provided by the specified traversable object.